src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -1249,3 +1238,49 @@ void dfcc_instrumentt::instrument_other(
                   << messaget::eom;
   }
 }
+
+void dfcc_instrumentt::apply_loop_contracts(
+  const irep_idt &function_id,
+  goto_functiont &goto_function,
+  dfcc_cfg_infot &cfg_info,
+  const dfcc_loop_contract_modet loop_contract_mode,
+  const std::set<symbol_exprt> &local_statics,
+  std::set<irep_idt> &function_pointer_contracts)
+{
+  PRECONDITION(loop_contract_mode != dfcc_loop_contract_modet::NONE);
+  cfg_info.get_loops_toposorted();
+
+  std::list<std::string> to_unwind;
+
+  // Apply loop contract transformations in topological order
+  for(const auto &loop_id : cfg_info.get_loops_toposorted())
+  {
+    const auto &loop = cfg_info.get_loop_info(loop_id);
+    if(loop.invariant.is_nil() && loop.decreases.empty())
+    {
+      // skip loops that do not have contracts
+      continue;
```

Comment:
It'd be nice to have debug information about skipped loops. Not a blocker.

Reply:
fixed